File: test/compilers/import_map_test.js
import { assertStringIncludes } from "testing/asserts.ts";
import { assertSnapshot } from "testing/snapshot.ts";
import { beforeEach, describe, it } from "testing/bdd.ts";
import { join } from "std/path/mod.ts";

import main from "../../lib/proscenium/compilers/esbuild.js";

const root = join(Deno.cwd(), "test", "dummy");
const lightningcssBin = join(Deno.cwd(), "bin", "lightningcss");

describe("import map", () => {
  beforeEach(() => {
    Deno.env.set("RAILS_ENV", "test");
    Deno.env.set("PROSCENIUM_TEST", "test");
  });

  it("map to a URL", async () => {
    const result = await main("lib/import_map/to_url.js", {
      root,
      lightningcssBin,
      importMap: "config/import_maps/to_url.json",
    });

    assertStringIncludes(
      new TextDecoder().decode(result),
      'import axios from "/url:https%3A%2F%2Fcdnjs.cloudflare.com%2Fajax%2Flibs%2Faxios%2F0.24.0%2Faxios.min.js";'
    );
  });

  it("from json", async () => {
    const result = await main("lib/import_map/bare_specifier.js", {
      root,
      lightningcssBin,
      importMap: "config/import_maps/simple.json",
    });

    const code = new TextDecoder().decode(result);

    assertStringIncludes(code, 'import foo from "/lib/foo.js";');
  });

  it("from js", async () => {
    const result = await main("lib/import_map/as_js.js", {
      root,
      lightningcssBin,
      importMap: "config/import_maps/as.js",
    });

    assertStringIncludes(
      new TextDecoder().decode(result),
      'import pkg from "/lib/foo2.js";'
    );
  });

  it("should map to bundle: prefix", async (t) => {
    const result = await main("lib/import_map/bare_modules.js", {
      root,
      lightningcssBin,
      importMap: "config/import_maps/bundle_prefix.json",
    });

    await assertSnapshot(t, new TextDecoder().decode(result));
  });

  it("should map to bundle-all: prefix", async (t) => {
    const result = await main("lib/import_map/bare_modules.js", {
      root,
      lightningcssBin,
      importMap: "config/import_maps/bundle_all_prefix.json",
    });

    await assertSnapshot(t, new TextDecoder().decode(result));
  });

  it("should use root as resolveDir", async (t) => {
    const result = await main("lib/import_map/nested/index.js", {
      root,
      lightningcssBin,
      importMap: "config/import_maps/external_imports/basic.json",
    });

    await assertSnapshot(t, new TextDecoder().decode(result));
  });

  it("should use root as resolveDir when using bundle: prefix", async (t) => {
    const result = await main("npm:@external/three/src/prop_types", {
      root,
      lightningcssBin,
      importMap: "config/import_maps/external_imports/bundle.json",
    });

    await assertSnapshot(t, new TextDecoder().decode(result));
  });

  it("should use root as resolveDir when using bundle-all: prefix", async (t) => {
    const result = await main("npm:@external/three/src/prop_types", {
      root,
      lightningcssBin,
      importMap: "config/import_maps/external_imports/bundle-all.json",
    });

    await assertSnapshot(t, new TextDecoder().decode(result));
  });

  it("maps imports via trailing slash", async () => {
    const result = await main("lib/component.jsx", {
      root,
      lightningcssBin,
      importMap: "config/import_maps/trailing_slash_import.json",
    });

    assertStringIncludes(
      new TextDecoder().decode(result),
      'import { jsx } from "/url:https%3A%2F%2Fesm.sh%2Freact%4018.2.0%2Fjsx-runtime"'
    );
  });

  it("resolves imports from a node_module", async () => {
    const result = await main("node_modules/is-ip/index.js", {
      root,
      lightningcssBin,
      importMap: "config/import_maps/npm.json",
    });

    assertStringIncludes(
      new TextDecoder().decode(result),
      'import ipRegex from "/url:https%3A%2F%2Fesm.sh%2Fip-regex";'
    );
  });

  it("supports scopes", async () => {
    const result = await main("lib/import_map/scopes.js", {
      root,
      lightningcssBin,
      importMap: "config/import_maps/scopes.json",
    });

    assertStringIncludes(
      new TextDecoder().decode(result),
      'import foo from "/lib/foo4.js";'
    );
  });

  it("should map bare modules", async () => {
    const result = await main("lib/import_map/bare_modules.js", {
      root,
      lightningcssBin,
      importMap: "config/import_maps/bare_modules.json",
    });

    assertStringIncludes(
      new TextDecoder().decode(result),
      'import { isIP } from "/node_modules/.pnpm/is-ip@5.0.0/node_modules/is-ip/index.js";'
    );
  });
});

File: test/import_map_test.go
package proscenium_test

import (
	b "joelmoss/proscenium/internal/builder"
	"joelmoss/proscenium/internal/importmap"
	"joelmoss/proscenium/internal/types"
	. "joelmoss/proscenium/test/support"
	"strings"

	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
)

var _ = Describe("import maps", func() {
	It("errors with invalid json", func() {
		result := importmap.NewJsonImportMap([]byte(`{`))

		Expect(strings.HasPrefix(result.Error(), "Cannot read import map:")).To(BeTrue())
	})

	It("populates build errors with invalid json", func() {
		result := importmap.NewJsonImportMap([]byte(`{`))

		Expect(strings.HasPrefix(result.Error(), "Cannot read import map:")).To(BeTrue())
	})

	When("import map is JS", func() {
		It("errors when invalid contents", func() {
			result := importmap.NewJavaScriptImportMap([]byte(`()(())`))

			Expect(strings.HasPrefix(result.Error(), "Cannot read import map:")).To(BeTrue())
		})

		It("populates build errors when invalid", func() {
			importmap.TEST_IMPORT_MAP_FILE = "config/import_maps/invalid.js"
			importmap.TEST_IMPORT_MAP_TYPE = importmap.JavascriptType

			result := importmap.NewJavaScriptImportMap([]byte(`()(())`))

			Expect(strings.HasPrefix(result.Error(), "Cannot read import map:")).To(BeTrue())
		})
	})

	Describe("bundle = true", func() {
		BeforeEach(func() {
			types.Config.Bundle = true
		})

		It("parses import map as JS", func() {
			importmap.NewJavaScriptImportMap([]byte(`
				env => ({
					imports: {
						pkg: env === 'test' ? '/lib/foo2.js' : '/lib/foo3.js'
					}
				})
			`))

			_, code, _ := b.BuildToString("lib/import_map/as_js.js")

			Expect(code).To(ContainCode(`console.log("/lib/foo2.js");`))
		})

		// import foo from 'foo'
		Describe("bare specifier", func() {
			When("value starts with /", func() {
				It("resolves", func() {
					importmap.NewJsonImportMap([]byte(`{
						"imports": { "foo": "/lib/foo.js" }
					}`))

					_, code, _ := b.BuildToString("lib/import_map/bare_specifier.js")

					Expect(code).To(ContainCode(`console.log("/lib/foo.js");`))
				})
			})

			When("value starts with ./ or ../", func() {
				It("resolves", func() {
					importmap.NewJsonImportMap([]byte(`{
						"imports": { "foo": "../foo.js" }
					}`))

					_, code, _ := b.BuildToString("lib/import_map/bare_specifier.js")

					Expect(code).To(ContainCode(`console.log("/lib/foo.js");`))
				})
			})

			When("value is URL", func() {
				It("is not bundled", func() {
					MockURL("/foo.js", "console.log('foo');")

					importmap.NewJsonImportMap([]byte(`{
						"imports": { "foo": "https://proscenium.test/foo.js" }
					}`))

					_, code, _ := b.BuildToString("lib/import_map/bare_specifier.js")

					Expect(code).To(ContainCode(`import "https://proscenium.test/foo.js";`))
				})
			})

			When("value is bare specifier", func() {
				It("resolves the value", func() {
					importmap.NewJsonImportMap([]byte(`{
						"imports": { "foo": "pkg" }
					}`))

					_, code, _ := b.BuildToString("lib/import_map/bare_specifier.js")

					Expect(code).To(ContainCode(`console.log("pkg/index.js")`))
				})
			})

			When("value is directory containing an index file", func() {
				It("resolves the value to index file", func() {
					importmap.NewJsonImportMap([]byte(`{
						"imports": { "foo": "/lib/indexes" }
					}`))

					_, code, _ := b.BuildToString("lib/import_map/bare_specifier.js")

					Expect(code).To(ContainCode(`
						console.log("lib/indexes/index.js");
					`))
				})
			})

			It("resolves file without extension", func() {
				importmap.NewJsonImportMap([]byte(`{
					"imports": { "foo": "/lib/foo2" }
				}`))

				_, code, _ := b.BuildToString("lib/import_map/bare_specifier.js")

				Expect(code).To(ContainCode(`console.log("/lib/foo2.js");`))
			})
		})

		// import foo from "foo/one.js"
		When("key and value have trailing slash", func() {
			It("resolves", func() {
				importmap.NewJsonImportMap([]byte(`{
					"imports": { "foo/": "./nested/foo/" }
			}`))

				_, code, _ := b.BuildToString("lib/import_map/path_prefix.js")

				Expect(code).To(ContainCode(`console.log("/lib/import_map/nested/foo/one.js");`))
			})
		})

		It("resolves to URL", func() {
			importmap.NewJsonImportMap([]byte(`{
				"imports": { "axios": "https://proscenium.test/axios.js" }
			}`))

			_, code, _ := b.BuildToString("lib/import_map/to_url.js")

			Expect(code).To(ContainCode(`
				import axios from "https://proscenium.test/axios.js";
			`))
		})

		It("resolves to bare module", func() {
			importmap.NewJsonImportMap([]byte(`{
				"imports": { "my-package": "pkg" }
			}`))

			_, code, _ := b.BuildToString("lib/import_map/bare_modules.js")

			Expect(code).To(ContainCode(`console.log("pkg/index.js");`))
		})

		XIt("scopes", Pending, func() {
			importmap.NewJsonImportMap([]byte(`{
				"imports": {
					"foo": "/lib/foo.js"
				},
				"scopes": {
					"/lib/import_map/": {
						"foo": "/lib/foo4.js"
					}
				}
			}`))

			_, code, _ := b.BuildToString("lib/import_map/scopes.js")

			Expect(code).To(ContainCode(`import foo from "/lib/foo4.js";`))
		})
	})

	Describe("bundle = false", func() {
		BeforeEach(func() {
			types.Config.Bundle = false
		})

		It("parses import map as JS", func() {
			importmap.NewJavaScriptImportMap([]byte(`
					env => ({
						imports: {
							pkg: env === 'test' ? '/lib/foo2.js' : '/lib/foo3.js'
						}
					})
				`))

			_, code, _ := b.BuildToString("lib/import_map/as_js.js")

			Expect(code).To(ContainCode(`import pkg from "/lib/foo2.js";`))
		})

		It("parse import map as JSON", func() {
			importmap.NewJsonImportMap([]byte(`{
				"imports": { "foo": "/lib/foo.js" }
			}`))

			_, code, _ := b.BuildToString("lib/import_map/bare_specifier.js")

			Expect(code).To(ContainCode(`import "/lib/foo.js";`))
		})

		// import foo from 'foo'
		Describe("bare specifier", func() {
			When("value starts with /", func() {
				It("resolves", func() {
					importmap.NewJsonImportMap([]byte(`{
						"imports": { "foo": "/lib/foo.js" }
					}`))

					_, code, _ := b.BuildToString("lib/import_map/bare_specifier.js")

					Expect(code).To(ContainCode(`import "/lib/foo.js";`))
				})
			})

			When("value starts with ./ or ../", func() {
				It("resolves", func() {
					importmap.NewJsonImportMap([]byte(`{
						"imports": { "foo": "../foo.js" }
					}`))

					_, code, _ := b.BuildToString("lib/import_map/bare_specifier.js")

					Expect(code).To(ContainCode(`import "/lib/foo.js";`))
				})
			})

			When("value is URL", func() {
				It("is not bundled", func() {
					MockURL("/foo.js", "console.log('foo');")

					importmap.NewJsonImportMap([]byte(`{
						"imports": { "foo": "https://proscenium.test/foo.js" }
					}`))

					_, code, _ := b.BuildToString("lib/import_map/bare_specifier.js")

					Expect(code).To(ContainCode(`import "https://proscenium.test/foo.js";`))
				})
			})

			When("value is bare specifier", func() {
				It("resolves the value", func() {
					importmap.NewJsonImportMap([]byte(`{
						"imports": { "foo": "pkg" }
					}`))

					_, code, _ := b.BuildToString("lib/import_map/bare_specifier.js")

					Expect(code).To(ContainCode(`import "/node_modules/pkg/index.js";`))
				})
			})

			When("value is directory containing an index file", func() {
				It("resolves the value to index file", func() {
					importmap.NewJsonImportMap([]byte(`{
						"imports": { "foo": "/lib/indexes" }
					}`))

					_, code, _ := b.BuildToString("lib/import_map/bare_specifier.js")

					Expect(code).To(ContainCode(`import "/lib/indexes/index.js";`))
				})
			})

			It("resolves file without extension", func() {
				importmap.NewJsonImportMap([]byte(`{
					"imports": { "foo": "/lib/foo2" }
				}`))

				_, code, _ := b.BuildToString("lib/import_map/bare_specifier.js")

				Expect(code).To(ContainCode(`import "/lib/foo2.js";`))
			})
		})

		// import foo from "foo/one.js"
		When("key and value have trailing slash", func() {
			It("resolves", func() {
				importmap.NewJsonImportMap([]byte(`{
					"imports": { "foo/": "./nested/foo/" }
			}`))

				_, code, _ := b.BuildToString("lib/import_map/path_prefix.js")

				Expect(code).To(ContainCode(`
					import foo from "/lib/import_map/nested/foo/one.js";
				`))
			})
		})

		It("resolves to URL", func() {
			importmap.NewJsonImportMap([]byte(`{
				"imports": { "axios": "https://proscenium.test/axios.js" }
			}`))

			_, code, _ := b.BuildToString("lib/import_map/to_url.js")

			Expect(code).To(ContainCode(`
				import axios from "https://proscenium.test/axios.js";
			`))
		})

		It("resolves to bare module", func() {
			importmap.NewJsonImportMap([]byte(`{
				"imports": { "my-package": "pkg" }
			}`))

			_, code, _ := b.BuildToString("lib/import_map/bare_modules.js")

			Expect(code).To(ContainCode(`import mypackage from "/node_modules/pkg/index.js";`))
		})
	})
})
